### ISSUE TITLE
Docker prestart hook breaks networks after Docker restart

### ISSUE THREAD

**Issue by stevenjm** [kubernetes/kops#6191]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

```
Version 1.10.0 (git-8b52ea6d1)
```

We are using this with our own `nodeup` build, which has a backport of https://github.com/kubernetes/kops/commit/6207b560778102d7208fdc68db7f8107ae21170e and the Docker prestart hook removed.

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"archive", BuildDate:"2018-12-04T23:49:50Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.11", GitCommit:"637c7e288581ee40ab4ca210618a89a555b6e7e9", GitTreeState:"clean", BuildDate:"2018-11-26T14:25:46Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

**3. What cloud provider are you using?**

AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

`systemctl restart docker.service` on a running node.

This means that it gets triggered if the Docker healthcheck fails, so it is tangentially related to #4969, but a Docker restart may also happen for other reasons and should not break the node in any case.

**5. What happened after the commands executed?**

```
# docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
# 
```

**6. What did you expect to happen?**

```
# docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
c2aa941fb292        bridge              bridge              local
f99cddc538f6        host                host                local
005e9341ea07        none                null                local
# 
```

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```
apiVersion: kops/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: null
  name: XXX
spec:
  additionalPolicies:
    master: |
      [
        {
          "Action": "ec2:DescribeTags",
          "Effect": "Allow",
          "Resource": "*"
        },
        {
          "Action": [
            "logs:CreateLogGroup",
            "logs:CreateLogStream",
            "logs:DescribeLogStreams",
            "logs:PutLogEvents"
          ],
          "Effect": "Allow",
          "Resource": [
            "arn:aws:logs:eu-west-1:XXX:log-group:/XXX/*"
          ]
        },
        {
          "Action": "sts:AssumeRole",
          "Effect": "Allow",
          "Resource": "arn:aws:iam::XXX:role/k8s/*"
        }
      ]
    node: |
      [
        {
          "Action": "ec2:CreateTags",
          "Effect": "Allow",
          "Resource": "arn:aws:ec2:eu-west-1::image/ami-*"
        },
        {
          "Action": "ec2:DescribeTags",
          "Effect": "Allow",
          "Resource": "*"
        },
        {
          "Action": [
            "logs:CreateLogGroup",
            "logs:CreateLogStream",
            "logs:DescribeLogStreams",
            "logs:PutLogEvents"
          ],
          "Effect": "Allow",
          "Resource": [
            "arn:aws:logs:eu-west-1:XXX:log-group:/XXX/*"
          ]
        }
      ]
  api:
    dns: {}
  authorization:
    rbac: {}
  channel: stable
  cloudLabels:
    Managed By: XXX
    cluster: XXX
  cloudProvider: aws
  configBase: s3://XXX
  dnsZone: Z277T5OMHP5I8U
  docker:
    version: 18.06.1
  etcdClusters:
  - enableEtcdTLS: true
    enableTLSAuth: true
    etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-eu-west-1a
      kmsKeyId: XXX
      name: eu-west-1a
    - encryptedVolume: true
      instanceGroup: master-eu-west-1b
      kmsKeyId: XXX
      name: eu-west-1b
    - encryptedVolume: true
      instanceGroup: master-eu-west-1c
      kmsKeyId: XXX
      name: eu-west-1c
    name: main
    version: 3.0.17
  - enableEtcdTLS: true
    enableTLSAuth: true
    etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-eu-west-1a
      name: eu-west-1a
    - encryptedVolume: true
      instanceGroup: master-eu-west-1b
      name: eu-west-1b
    - encryptedVolume: true
      instanceGroup: master-eu-west-1c
      name: eu-west-1c
    name: events
    version: 3.0.17
  fileAssets:
  - content: XXX-master
    name: servername
    path: /etc/awslogs/servername
    roles:
    - Master
  - content: XXX-node
    name: servername
    path: /etc/awslogs/servername
    roles:
    - Node
  - content: "{\n\t\"cniVersion\": \"0.3.0\",\n\t\"name\": \"mynet\",\n\t\"plugins\":
      [\n\t\t{\n\t\t\t\"name\": \"weave\",\n\t\t\t\"type\": \"weave-net\",\n\t\t\t\"hairpinMode\":
      true\n\t\t},\n\t\t{\n\t\t\t\"type\": \"portmap\",\n\t\t\t\"capabilities\": {\"portMappings\":
      true},\n\t\t\t\"snat\": true\n\t\t}\n\t]\n}\n"
    name: 00-weave.conflist
    path: /etc/cni/net.d/00-weave.conflist
  - content: |
      apiVersion: audit.k8s.io/v1beta1  #this is required in Kubernetes 1.8
      kind: Policy
      rules:
        # Drop all read requests
        - level: None
          verbs: ["watch", "get", "list"]
        # node and pod status calls from nodes are high-volume and can be large, don't log responses for expected updates from nodes
        - level: None
          users: ["kubelet", "system:node-problem-detector", "system:serviceaccount:kube-system:node-problem-detector"]
          verbs: ["update","patch"]
          resources:
            - group: "" # core
              resources: ["nodes/status", "pods/status", "apiservices/status"]
          omitStages:
            - "RequestReceived"
        - level: None
          userGroups: ["system:nodes"]
          verbs: ["update","patch"]
          resources:
            - group: "" # core
              resources: ["nodes/status", "pods/status", "apiservices/status"]
          omitStages:
            - "RequestReceived"
        - level: None
          verbs: ["update"]
          sourceIPs:
            - 127.0.0.1
          users: ["system:kube-scheduler", "system:kube-controller-manager"]
          resources:
            - group: ""
              resources: ["endpoints"]
        # deletecollection calls can be large, don't log responses for expected namespace deletions
        - level: Request
          users: ["system:serviceaccount:kube-system:namespace-controller"]
          verbs: ["deletecollection"]
          omitStages:
            - "RequestReceived"
        # Secrets, ConfigMaps, and TokenReviews can contain sensitive & binary data,
        # so only log at the Metadata level.
        - level: Metadata
          resources:
            - group: "" # core
              resources: ["secrets", "configmaps"]
            - group: authentication.k8s.io
              resources: ["tokenreviews"]
          omitStages:
            - "RequestReceived"
        - level: Request
          verbs: ["watch", "list", "get"]
          sourceIPs:
            - 10.*
          omitStages:
            - "RequestReceived"
        # kops repeatedly tries to create kube-dns, kube-proxy etc resources.
        - level: None
          verbs: ["create", "update"]
          users: ["kops"]
          sourceIPs:
            - 127.0.0.1
        - level: Request
          verbs: ["update", "create", "delete", "deletecollection"]
          omitStages:
            - "RequestReceived"
        # Default level for all other requests.
        - level: Metadata
          omitStages:
            - "RequestReceived"
    name: audit-policy.conf
    path: /srv/kubernetes/audit-policy.conf
    roles:
    - Master
  - content: |
      {
          "live-restore": true
      }
    name: docker-daemon.json
    path: /etc/docker/daemon.json
  - content: |
      XXX
    name: tag-ami
    path: /usr/local/sbin/tag-ami
  hooks:
  - manifest: |
      Type=oneshot
      RemainAfterExit=yes
      ExecStart=/bin/sh -c 'cp -n /var/cache/nodeup/extracted/*_https___storage_googleapis_com_kubernetes-release_network-plugins_cni-plugins-*/portmap /opt/cni/bin/'
    name: copy-portmap.service
  - manifest: |
      Type=oneshot
      RemainAfterExit=yes
      ExecStop=/bin/sh -c 'docker stop $(docker ps -q)'
    name: docker-stop.service
  - before:
    - docker-stop.service
    manifest: |
      Type=oneshot
      RemainAfterExit=yes
      ExecStart=/bin/sh -c 'perl -pi -e '\''s/^(Description=.*)$/$1\nAfter=docker.service/'\'' /lib/systemd/system/docker-stop.service && systemctl daemon-reload'
    name: docker-stop-fix.service
  - manifest: |
      Type=oneshot
      RemainAfterExit=yes
      ExecStart=/bin/sh -c 'chmod u+x /usr/local/sbin/tag-ami && exec /usr/local/sbin/tag-ami'
    name: tag-ami.service
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeAPIServer:
    auditLogMaxAge: 10
    auditLogMaxBackups: 1
    auditLogMaxSize: 100
    auditLogPath: /var/log/kube-apiserver-audit.log
    auditPolicyFile: /srv/kubernetes/audit-policy.conf
  kubernetesApiAccess:
  - 192.168.0.0/20
  - 10.0.0.0/14
  kubernetesVersion: 1.10.11
  masterInternalName: XXX
  masterPublicName: XXX
  networkCIDR: 10.3.0.0/16
  networkID: vpc-702e9215
  networking:
    weave: {}
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 192.168.0.0/20
  - 10.0.0.0/14
  sshKeyName: XXX
  subnets:
  - cidr: 10.3.0.0/19
    id: subnet-c3c7c0a4
    name: eu-west-1a
    type: Private
    zone: eu-west-1a
  - cidr: 10.3.32.0/19
    id: subnet-0b8d8542
    name: eu-west-1b
    type: Private
    zone: eu-west-1b
  - cidr: 10.3.64.0/19
    id: subnet-f5e4a4ae
    name: eu-west-1c
    type: Private
    zone: eu-west-1c
  - cidr: 10.3.248.0/24
    id: subnet-363c3b51
    name: utility-eu-west-1a
    type: Utility
    zone: eu-west-1a
  - cidr: 10.3.249.0/24
    id: subnet-a98189e0
    name: utility-eu-west-1b
    type: Utility
    zone: eu-west-1b
  - cidr: 10.3.250.0/24
    id: subnet-7ee3a325
    name: utility-eu-west-1c
    type: Utility
    zone: eu-west-1c
  topology:
    dns:
      type: Private
    masters: private
    nodes: private

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-12-05T15:38:16Z
  labels:
    kops.k8s.io/cluster: XXX
  name: canary
spec:
  additionalSecurityGroups:
  - sg-e72b4f9a
  associatePublicIp: false
  cloudLabels:
    XXX__Prometheus: node
  image: ami-08215bdfd498fe066
  kubelet:
    enforceNodeAllocatable: pods
    kubeReservedCgroup: /kube.slice
    kubeletCgroups: /kube.slice/kubelet
    runtimeCgroups: /kube.slice/runtime
    systemReserved:
      cpu: 100m
      memory: 400Mi
    systemReservedCgroup: /system.slice
  machineType: m4.large
  maxSize: 1
  minSize: 1
  nodeLabels:
    beta.kubernetes.io/fluentd-ds-ready: "true"
    kops.k8s.io/instancegroup: canary
    XXX/canary: "true"
  role: Node
  rootVolumeSize: 32
  subnets:
  - eu-west-1a
  - eu-west-1b
  - eu-west-1c

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-12-05T15:38:17Z
  labels:
    kops.k8s.io/cluster: XXX
  name: ingress
spec:
  additionalSecurityGroups:
  - sg-e72b4f9a
  associatePublicIp: false
  cloudLabels:
    XXX__Prometheus: node
  image: ami-08215bdfd498fe066
  kubelet:
    enforceNodeAllocatable: pods
    kubeReservedCgroup: /kube.slice
    kubeletCgroups: /kube.slice/kubelet
    runtimeCgroups: /kube.slice/runtime
    systemReserved:
      cpu: 100m
      memory: 400Mi
    systemReservedCgroup: /system.slice
  machineType: t2.large
  maxSize: 3
  minSize: 3
  nodeLabels:
    beta.kubernetes.io/fluentd-ds-ready: "true"
    kops.k8s.io/instancegroup: ingress
    layer: ingress
  role: Node
  subnets:
  - eu-west-1a
  - eu-west-1b
  - eu-west-1c
  taints:
  - key=ingress:NoSchedule

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-12-05T15:38:16Z
  labels:
    kops.k8s.io/cluster: XXX
  name: master-eu-west-1a
spec:
  additionalSecurityGroups:
  - sg-e72b4f9a
  associatePublicIp: false
  cloudLabels:
    XXX__Prometheus: node
  image: ami-08215bdfd498fe066
  kubelet:
    enforceNodeAllocatable: pods
    kubeReservedCgroup: /kube.slice
    kubeletCgroups: /kube.slice/kubelet
    runtimeCgroups: /kube.slice/runtime
    systemReserved:
      cpu: 100m
      memory: 400Mi
    systemReservedCgroup: /system.slice
  machineType: m4.large
  maxSize: 1
  minSize: 1
  nodeLabels:
    beta.kubernetes.io/fluentd-ds-ready: "true"
    kops.k8s.io/instancegroup: master-eu-west-1a
  role: Master
  subnets:
  - eu-west-1a

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-12-05T15:38:16Z
  labels:
    kops.k8s.io/cluster: XXX
  name: master-eu-west-1b
spec:
  additionalSecurityGroups:
  - sg-e72b4f9a
  associatePublicIp: false
  cloudLabels:
    XXX__Prometheus: node
  image: ami-08215bdfd498fe066
  kubelet:
    enforceNodeAllocatable: pods
    kubeReservedCgroup: /kube.slice
    kubeletCgroups: /kube.slice/kubelet
    runtimeCgroups: /kube.slice/runtime
    systemReserved:
      cpu: 100m
      memory: 400Mi
    systemReservedCgroup: /system.slice
  machineType: m4.large
  maxSize: 1
  minSize: 1
  nodeLabels:
    beta.kubernetes.io/fluentd-ds-ready: "true"
    kops.k8s.io/instancegroup: master-eu-west-1b
  role: Master
  subnets:
  - eu-west-1b

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-12-05T15:38:16Z
  labels:
    kops.k8s.io/cluster: XXX
  name: master-eu-west-1c
spec:
  additionalSecurityGroups:
  - sg-e72b4f9a
  associatePublicIp: false
  cloudLabels:
    XXX__Prometheus: node
  image: ami-08215bdfd498fe066
  kubelet:
    enforceNodeAllocatable: pods
    kubeReservedCgroup: /kube.slice
    kubeletCgroups: /kube.slice/kubelet
    runtimeCgroups: /kube.slice/runtime
    systemReserved:
      cpu: 100m
      memory: 400Mi
    systemReservedCgroup: /system.slice
  machineType: m4.large
  maxSize: 1
  minSize: 1
  nodeLabels:
    beta.kubernetes.io/fluentd-ds-ready: "true"
    kops.k8s.io/instancegroup: master-eu-west-1c
  role: Master
  subnets:
  - eu-west-1c

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-12-05T15:38:17Z
  labels:
    kops.k8s.io/cluster: XXX
  name: nodes
spec:
  additionalSecurityGroups:
  - sg-e72b4f9a
  associatePublicIp: false
  cloudLabels:
    XXX__Prometheus: node
  image: ami-08215bdfd498fe066
  kubelet:
    enforceNodeAllocatable: pods
    kubeReservedCgroup: /kube.slice
    kubeletCgroups: /kube.slice/kubelet
    runtimeCgroups: /kube.slice/runtime
    systemReserved:
      cpu: 100m
      memory: 400Mi
    systemReservedCgroup: /system.slice
  machineType: m4.large
  maxSize: 6
  minSize: 6
  nodeLabels:
    beta.kubernetes.io/fluentd-ds-ready: "true"
    kops.k8s.io/instancegroup: nodes
  role: Node
  subnets:
  - eu-west-1a
  - eu-west-1b
  - eu-west-1c
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

**9. Anything else do we need to know?**

Although (as you can see from our cluster manifest) we are running Docker with the `live-restore` setting, I tested this behaviour both with and without that setting, with the same result.

It looks like this was probably broken after [this change](https://github.com/moby/moby/commit/ecffb6d58cf89371e3f4a20f55c2e614dbdfe880#diff-b56258ecab11f7b797306907327cc904R641) in Docker, which caused network initialisation not to always occur on startup. This change postdates [the introduction](https://github.com/kubernetes/kops/commit/d4c2cfaae7f900586d87bd468042aa0b112b7c89) of the `docker-prestart` hook to Kops. I believe that the bug is in Kops because Kops is modifying Docker's internal state.

It's not clear whether [the issue](https://github.com/docker/docker/issues/18283) this hook is working around is still a problem, but I suspect the commit that broke Kops in this case will also fix this issue, since initialising the network interfaces is now conditional.

We have an existing patch we're using which just removes this hook. If this is considered the correct fix, let me know and I'll submit a PR with it.

**Comment by akshaychitneni** [comment]
We ran into same issue when we enabled live-restore.

**Comment by stevenjm** [comment]
Fixed in #6564.
